### PR TITLE
Add deform_conv2d as implemented in torchvision

### PR DESCRIPTION
**Pull request by wingertge**
## Add Deformable Convolution Layers v1/v2

Help would be welcome, I've been looking at the kernel so long I can no longer see my mistakes. Getting a few more pairs of eyes on it would be helpful.

### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

A lot of resnet variants currently replace most conv2d layers with deform_conv2d. See [torchvision docs](https://pytorch.org/vision/main/generated/torchvision.ops.deform_conv2d.html)

### Changes

Adds the relatively popular deform_conv2d layer initially implemented in torchvision in 2021.

### Testing

* Tested with comparison to pytorch output
* All tests passing

### CODE REVIEW

**Comment by wingertge** [comment]
Currently blocked on atomics being added to `cubecl`

**Comment by wingertge** [comment]
Finally done and ready to review

**Review by louisfd — CHANGES_REQUESTED** [review]
Just a few questions, otherwise very nice!

File: crates/burn-jit/src/kernel/conv/deform_conv2d.rs [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+            let interpolated = bilinear_interpolate(input, height, width, y, x, input_base_idx);
+
+            columns[col_base_idx] = mask_value * interpolated;
+            col_base_idx += col_stride_0;
+        }
+    }
+}
+
+#[cube]
+pub(crate) fn bilinear_interpolate<F: Float>(
```
[the remark this answers is not in the document]
Comment (wingertge) [reply crates/burn-jit/src/kernel/conv/deform_conv2d.rs:127]: I don't believe there is a cube function that does bilinear interpolate for a single element? All I could find was the kernel that does it for a whole tensor.

File: crates/burn-jit/src/kernel/conv/deform_conv_transpose2d.rs [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+                float_atomic_add::<F>(&mut grad_input[gradient_pos], value);
+            }
+        }
+    }
+}
+
+#[cube]
+fn float_atomic_add<F: Float>(ptr: &mut AtomicU32, value: F) {
+    if value == F::new(0.0) {
+        return;
```
[the remark this answers is not in the document]
Comment (wingertge) [reply crates/burn-jit/src/kernel/conv/deform_conv_transpose2d.rs:579]: I did forget about that.

File: crates/burn-jit/src/kernel/conv/deform_conv2d.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (louisfd) [reply crates/burn-jit/src/kernel/conv/deform_conv2d.rs:127]: Good, only wanted to check if you had realized something similar existed

**Review by louisfd — APPROVED** [review]
LGTM

**Pull request merged**